tree/tree.go

```diff
@@ -0,0 +1,723 @@
+	HelpStyle             lipgloss.Style
+	TreeStyle             lipgloss.Style
+	selectedNodeFunc      StyleFunc
+	SelectedNodeStyle     lipgloss.Style
+	SelectedNodeStyleFunc StyleFunc
+	nodeFunc              StyleFunc
+	NodeStyle             lipgloss.Style
+	NodeStyleFunc         StyleFunc
+}
+
+// DefaultStyles returns a set of default style definitions for this tree
+// component.
+func DefaultStyles() (s Styles) {
+	s.NodeStyle = lipgloss.NewStyle()
+	s.SelectedNodeStyle = lipgloss.NewStyle().
+		Background(lipgloss.Color("62")).
+		Foreground(lipgloss.Color("18")).
+		Bold(true)
+	s.nodeFunc = func(_ Nodes, _ int) lipgloss.Style {
+		return s.NodeStyle
+	}
+	s.selectedNodeFunc = func(_ Nodes, _ int) lipgloss.Style {
+		return s.SelectedNodeStyle
+	}
+	s.SelectionCursor = lipgloss.NewStyle().Foreground(lipgloss.Color("9"))
+	s.HelpStyle = lipgloss.NewStyle().PaddingTop(1)
+
+	return s
+}
+
+// KeyMap is the key bindings for different actions within the tree.
+type KeyMap struct {
+	Down   key.Binding
+	Up     key.Binding
+	Toggle key.Binding
+
+	// Help toggle keybindings.
+	ShowFullHelp  key.Binding
+	CloseFullHelp key.Binding
+
+	Quit key.Binding
+}
+
+// DefaultKeyMap is the default set of key bindings for navigating and acting
+// upon the tree.
+var DefaultKeyMap = KeyMap{
+	Down: key.NewBinding(
+		key.WithKeys("down", "j", "ctrl+n"),
+		key.WithHelp("↓/j", "down"),
+	),
+	Up: key.NewBinding(
+		key.WithKeys("up", "k", "ctrl+p"),
+		key.WithHelp("↑/k", "up"),
+	),
+	Toggle: key.NewBinding(
+		key.WithKeys("enter"),
+		key.WithHelp("⏎", "toggle"),
+	),
+
+	// Toggle help.
+	ShowFullHelp: key.NewBinding(
+		key.WithKeys("?"),
+		key.WithHelp("?", "more"),
+	),
+	CloseFullHelp: key.NewBinding(
+		key.WithKeys("?"),
+		key.WithHelp("?", "close help"),
+	),
+
+	Quit: key.NewBinding(
+		key.WithKeys("q", "ctrl+c"),
+		key.WithHelp("q", "quit"),
+	),
+}
+
+// Model is the Bubble Tea model for this tree element.
+type Model struct {
+	showHelp bool
+	// OpenCharacter is the character used to represent an open node.
+	OpenCharacter string
+	// ClosedCharacter is the character used to represent a closed node.
+	ClosedCharacter string
+	// KeyMap encodes the keybindings recognized by the widget.
+	KeyMap KeyMap
+	// styles sets the styling for the tree
+	styles Styles
+	Help   help.Model
+
+	// Additional key mappings for the short and full help views. This allows
+	// you to add additional key mappings to the help menu without
+	// re-implementing the help component. Of course, you can also disable the
+	// list's help component and implement a new one if you need more
+	// flexibility.
+	AdditionalShortHelpKeys func() []key.Binding
+	AdditionalFullHelpKeys  func() []key.Binding
+
+	root   *Node
+	width  int
+	height int
+	// yOffset is the vertical offset of the selected node.
+	yOffset int
+}
+
+// Node is a a node in the tree
+// Node implements lipgloss's tree.Node
+type Node struct {
+	// tree is used as the renderer layer
+	tree *ltree.Tree
```

Comment:
I'm thinking whether I should implement everything here, and not use `lipgloss.Tree` because:
1. I'm using lipgloss.Tree as a renderer, but I found myself implementing a lot of its logic.
2. I've also found that the `table` bubble doesn't actually use `lipgloss.Table`
3. I needed more granular styling like the notion of a selected node, which `lipgloss.Tree` doesn't "natively" support, so I wasn't sure if what I'm doing is an abuse.
4. I would also need to expose everything `lipgloss.Tree` exports like all the different Enumerator styles and the rest.

Let me know what you think. Should the view function also be implemented here and not use `lipgloss.Tree.View()`?

Reply:
Ideally, we use `lipgloss/tree` for rendering. If it doesn't support what we need we can get it there.

So for some background, the `table/bubble` predates `lipgloss.Table`, but the plan was always to migrate to Lip Gloss for rendering as `lipgloss/table`, which is much, much more robust than the Bubbles implementation.